src/DescriptorsCommand.cc

```diff
@@ -409,94 +412,240 @@ void AddDescriptor::retrieve_aws_descriptorSet(const std::string &set_path) {
+        std::cerr << blob.length() / 4 << " " << dim << std::endl;
+        error["info"] = "Blob Dimensions Mismatch";
+        return -1;
+    }
+
+    // retrieve the descriptor set from AWS here
+    // operations are currently done in memory with no subsequent write to disk
+    // so there's no need to re-upload to AWS
+    if (_use_aws_storage) {
+        retrieve_aws_descriptorSet(set_path);
+    }
 
-  long id = insert_descriptor(blob, set_path, dimensions, label, error);
+    //TODO modify insert descriptor to handle batches
+    long id = insert_descriptor(blob, set_path, 1, label, error);
 
-  if (id < 0) {
-    error["status"] = RSCommand::Error;
+    if (id < 0) {
+        error["status"] = RSCommand::Error;
+
+        if (_use_aws_storage) {
+            // delete files in set_path
+            std::uintmax_t n = fs::remove_all(set_path);
+            std::cout << "Deleted " << n << " files or directories\n";
+        }
+
+        return -1;
+    }
+
+    props[VDMS_DESC_ID_PROP] = Json::Int64(id);
+
+    int node_ref = get_value<int>(cmd, "_ref", query.get_available_reference());
+
+    query.AddNode(node_ref, VDMS_DESC_TAG, props, Json::nullValue);
+
+    // It passed the checker, so it exists.
+    int set_ref = query.get_available_reference();
+
+    Json::Value link;
+    Json::Value results;
+    Json::Value list_arr;
+    list_arr.append(VDMS_DESC_SET_PATH_PROP);
+    list_arr.append(VDMS_DESC_SET_DIM_PROP);
+    results["list"] = list_arr;
+
+    Json::Value constraints;
+    Json::Value name_arr;
+    name_arr.append("==");
+    name_arr.append(set_name);
+    constraints[VDMS_DESC_SET_NAME_PROP] = name_arr;
+
+    bool unique = true;
+
+    // Query set node
+    query.QueryNode(set_ref, VDMS_DESC_SET_TAG, link, constraints, results,
+                    unique);
+
+    if (cmd.isMember("link")) {
+        add_link(query, cmd["link"], node_ref, VDMS_DESC_EDGE_TAG);
+    }
+
+    Json::Value props_edge;
+    query.AddEdge(-1, set_ref, node_ref, VDMS_DESC_SET_EDGE_TAG, props_edge);
+
+    // TODO: deleting files here causes problems with concurrency (TestRetail.py)
+    // keeping local copies as a temporary solution
+    // if(_use_aws_storage)
+    // {
+    //     //delete files in set_path
+    //     std::uintmax_t n = fs::remove_all(set_path);
+    //     std::cout << "Deleted " << n << " files or directories\n";
+    // }
+
+    return 0;
 
+}
+
+int AddDescriptor::add_descriptor_batch(PMGDQuery &query,
+                                         const Json::Value &jsoncmd,
+                                         const std::string &blob, int grp_id,
+                                         Json::Value &error){
+
+    int expected_blb_size;
+    int nr_expected_descs;
+    int dimensions;
+
+    //Extract set name
+    const Json::Value &cmd = jsoncmd[_cmd_name];
+    const std::string set_name = cmd["set"].asString();
+
+    Json::Value props = get_value<Json::Value>(cmd, "properties");
```

Comment:
can an insertDescriptor query contain both 'properties' and 'batch_properties'?
If not, then some warning/error response when both are seen is probably needed.

If yes, then the 'properties' likely contains common properties to be applied for each descriptor in the batch. In this case, the props extracted here should be added into cur_props around line 586-587.

Reply:
Good catch, technically yes it can, but properties is ignored in favor of batch properties: I can just remove that code here.